src/AtomicRingBuffer.h

```diff
@@ -0,0 +1,78 @@
+#ifndef _ATOMIC_RING_BUFFER_H_
+#define _ATOMIC_RING_BUFFER_H_
+
+#include <atomic>
+#include <cassert>
+
+
+template <typename T>
+class AtomicRingBuffer
+{
+public:
+    AtomicRingBuffer(int capacity) : 
+        mBufferLength(capacity)
+        {
+            
+            assert(capacity > 0);
+            mBuffer = new T[capacity];
+        }
+    ~AtomicRingBuffer()
+    {
+        delete[] mBuffer;
+    }
+    
+    bool push(const T& element)
+    {
+        size_t currentWritePointer = mWritePointer.load();
+        size_t nextWritePointer = increment(currentWritePointer);
+        if (nextWritePointer != mReadPointer.load())
```

Comment:
Okay, I see that this does check for overwriting before storing the element, leaving it up to the caller to decide the appropriate response. (Depending on the importance of the thing being pushed, the correct answer might be 'throw an exception'.)

Reply:
Gotcha. For now I'll put an assert there that will trigger if something failed to be pushed. If this assert is easily triggered, I'll find a way around it.